src/choreo/test_choreo.c

```diff
@@ -0,0 +1,1163 @@
+    FD_TEST( query->ele );
+    parent_funk_txn = query->ele;
+  }
+
+  fd_funk_txn_xid_t xid;
+  xid.ul[0] = xid.ul[1] = entry.slot;
+  funk_txn = fd_funk_txn_prepare( funk, parent_funk_txn, &xid, 1 );
+
+  FD_TEST( funk_txn );
+  return funk_txn;
+}
+
+void
+setup_funk_txns( fd_funk_t * funk, ghost_entry_t * ghost_entries, ulong ghost_len, fd_pubkey_t * voter, fd_pubkey_t * identity, ulong voter_cnt, fd_tower_t * towers[] ) {
+  for ( ulong g=0; g<ghost_len; g++ ) {
+    bool is_root = g==0UL ? true : false;
+    fd_funk_txn_t * ghost_slot_funk_txn = mock_funk_txn( ghost_entries[g], is_root, funk );
+    for ( ulong v=0; v<voter_cnt; v++ ) {
+      ulong current_tower_height = fd_tower_votes_cnt( towers[v] );
+      for ( ulong t=0; t<current_tower_height; t++ ) {
+        if ( fd_tower_votes_peek_index_const( towers[v], t )->slot==ghost_entries[g].slot ) {
+          setup_voter_in_funk_txn( funk, ghost_slot_funk_txn, &voter[v], &identity[v] );
+        } else if ( fd_tower_votes_peek_index_const( towers[v], t )->slot>ghost_entries[g].slot ) {
+          break;
+        }
+      }
+    }
+  }
+}
+
+/*
+VOTE_FOR_SLOT( vote_slot, slot, voter_idx )
+ - vote_slot: slot to vote for
+ - slot: slot to vote in
+ - voter_idx: index of the voter in the voter array
+*/
+#define VOTE_FOR_SLOT( vote_slot, slot_num, voter_num ) \
+  fd_tower_t * vote_tower_##slot_num##_##voter_num = towers[voter_num]; \
+  ulong tower_height_##slot_num##_##voter_num = fd_tower_votes_cnt( vote_tower_##slot_num##_##voter_num ); \
+  ulong vote_slot_idx_##slot_num##_##voter_num = ULONG_MAX; \
+  for (ulong i = 0; i < tower_height_##slot_num##_##voter_num; i++) { \
+    if (fd_tower_votes_peek_index_const( vote_tower_##slot_num##_##voter_num, i )->slot == vote_slot) { \
+      vote_slot_idx_##slot_num##_##voter_num = i; \
+      break; \
+    } \
+  } \
+  if (vote_slot_idx_##slot_num##_##voter_num == ULONG_MAX) { \
+    FD_LOG_ERR( ( "VOTE_FOR_SLOT: vote_slot=%lu not found in tower", vote_slot ) ); \
+    return; \
+  } \
+  fd_landed_vote_t * landed_votes_##slot_num##_##voter_num = mock_landed_votes( wksp, tower_height_##slot_num##_##voter_num, vote_tower_##slot_num##_##voter_num, (uint)vote_slot_idx_##slot_num##_##voter_num ); \
+  fd_funk_txn_xid_t xid_##slot_num##_##voter_num; \
+  xid_##slot_num##_##voter_num.ul[0] = xid_##slot_num##_##voter_num.ul[1] = slot_num; \
+  fd_funk_txn_map_query_t query_##slot_num##_##voter_num[1]; \
+  fd_funk_txn_map_t txn_map_##slot_num##_##voter_num = fd_funk_txn_map( funk, fd_funk_wksp( funk ) ); \
+  FD_TEST( fd_funk_txn_map_query_try( &txn_map_##slot_num##_##voter_num, &xid_##slot_num##_##voter_num, NULL, query_##slot_num##_##voter_num, 0 )==FD_MAP_SUCCESS ); \
+  FD_TEST( query_##slot_num##_##voter_num->ele ); \
+  fd_funk_txn_t * funk_txn_##slot_num##_##voter_num = query_##slot_num##_##voter_num->ele; \
+  set_vote_account_tower( wksp, funk, funk_txn_##slot_num##_##voter_num, &voter[voter_num], landed_votes_##slot_num##_##voter_num );
```

Comment:
i get what you're addressing my feedback about helper macros, but this one is hard to read. a cleaner way to do this is to write a generic conversion function `tower_to_vote_state`. you can see `fd_tower_from_vote_acc` to see how it converts the other way

and so you always setup everything in terms of `TOWER(1, 2, 3, 4)` etc. then you call that helper to get a `fd_vote_state_t`.

you'll still need a helper function to put it in `funk`